File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool set[1001] = {};
    void dfs(TreeNode* &root, bool flag, vector<TreeNode*>& result){
        if (root == nullptr) 
            return;
        dfs(root->left, set[root->val], result);
        dfs(root->right, set[root->val], result);
        if (!set[root->val] && flag) result.push_back(root);
        if(set[root->val]) root = nullptr;
    }
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> result;
        for(int num : to_delete)
            set[num] = true;
        dfs(root, true, result);
        return result;
    }
};

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def delNodes(self, root, to_delete):
        """
        :type root: TreeNode
        :type to_delete: List[int]
        :rtype: List[TreeNode]
        """
        
        # BFS, O(n) time, O(n) space
        
        s = set(to_delete)
        result = []
        
        def dfs(root, flag):
            if not root:
                return None
            toDelete = (root.val in s)
            root.left = dfs(root.left, toDelete)
            root.right = dfs(root.right, toDelete)
            if not toDelete and flag:
                result.append(root)
            return None if toDelete else root
        
        dfs(root, True)
        return result
        